projects/memcached-clusters/cache-tester/cmd/root.go

```diff
@@ -0,0 +1,71 @@
+package cmd
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/bradfitz/gomemcache/memcache"
+)
+
+func Execute() {
+	mcrouter := flag.String("mcrouter", "", "port to connect to mcrouter")
+	memcacheds := flag.String("memcacheds", "", "comma separated list of memcached servers ports")
+	flag.Parse()
+
+	if *mcrouter == "" || *memcacheds == "" {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	// Connect to mcrouter
+	mcr := memcache.New("localhost:" + *mcrouter)
+	defer mcr.Close()
+
+	// Connect to memcached servers
+	ports := strings.Split(*memcacheds, ",")
+	mcs := make([]*memcache.Client, 0)
+	for i, port := range ports {
+		mcs = append(mcs, memcache.New("localhost:"+port))
+		defer mcs[i].Close()
+	}
+
+	// Set a key in mcrouter
+	err := mcr.Set(&memcache.Item{Key: "hello", Value: []byte("world")})
+	if err != nil {
+		fmt.Printf("Error setting key in mcrouter: %v\n", err)
+		os.Exit(1)
+	}
+
+	// Get the key from each memcached server
+	items := make([]*memcache.Item, 0)
+	item := new(memcache.Item)
+	for i, mc := range mcs {
+		item, err = mc.Get("hello")
+		if err != nil {
+			fmt.Printf("Error getting key from memcached server %v: %v\n", i, err)
+		}
+		items = append(items, item)
+
+	}
+
+	if err == memcache.ErrCacheMiss && len(items) != 0 {
```

Comment:
What exactly is the signal you're looking for that indicates that the cluster is running sharded?

This looks at the last error to be returned (rather than whether _any_ or _most_ or _all but one_ returned `memcache.ErrCacheMiss`), and AFAICT `items` will _always_ have length of the number of memcache servers present (because you always push to it in the loop).

Reply:
I don't remember what was my logic for using `len(items) != 0` -_-
I updated the code to include the err check inside the loop. so if we have any one returned `memcache.ErrCacheMiss` the mcrouter considered sharded .